### ISSUE TITLE
Type narrowing fails for discriminated unions when they occur as a type argument in a type alias

### ISSUE THREAD

**Issue by a1650078** [Microsoft/TypeScript#11185]
**TypeScript Version:**  Version 2.1.0-dev.20160926

**Code**

*Disclaimer: it's after 3am here*

In the following code, the type guard on Shape works as expected - but the inner if block doesn't narrow the type of s to circle (which would be the normal behaviour for discriminated unions). The alternative definition for Shapes works fine though. After trying some different combinations, it seems this only happens when the type alias uses Shape as a type argument for a generic type.

```ts
class Square { kind:"square"; }
class Circle { kind:"circle"; }

type Shape = Circle | Square;
type Shapes = Shape | Array<Shape>;  // <- this doesn't work
//type Shapes = Shape | string;      // <- this works fine

function isShape(s : Shapes): s is Shape {
    return !Array.isArray(s);
}

function fail(s: Shapes){
    if(isShape(s)){
        if(s.kind === "circle"){ // <- Won't narrow type
        // if(isCircle(s)){      // <- Custom type guard still works
            let c:Circle = s;
        }
    }
}
```

**Expected behavior:**

Compiles

**Actual behavior:** 
```
error TS2322: Type 'Shape' is not assignable to type 'Circle'.
  Type 'Square' is not assignable to type 'Circle'.
    Types of property 'kind' are incompatible.
      Type '"square"' is not assignable to type '"circle"'.
```

**Comment by ahejlsberg** [comment]
This is effectively the same issue as #10586.  We only consider a property to be a discriminant property if it is present in all constituents of the declared type, which isn't the case with the `kind` property when the declared type is `Shapes` (because `Shape[]` doesn't have a `kind`).

It seems we need to be a little more flexible here. A possible solution might be to consider a property to be a discriminant property if it occurs in more than one constituent (but not necessarily all constituents). I will see if I can come up with a fix.

**Comment by a1650078** [comment]
Hmm - at first I thought it was more complicated than that because `string` doesn't have a kind either, yet it compiles with `type Shapes = Shape | string`. But I guess the compiler ignores primitives when looking for discriminants.

If it really is too hard to make this work with the control flow analyser, maybe the compiler should keep track of aliases with discriminants when they are used in custom type guards or similar. It really only needs to work as well as function declarations, so that users don't extrapolate from the one context to the other.

```ts
class Square { kind:"square"; }
class Circle { kind:"circle"; }

type Shape = Circle | Square;
type Shapes = Shape | Array<Shape>;

function isShape(s : Shapes): s is Shape {
    return !Array.isArray(s);
}

function succeed(s : Shape){
    if(s.kind === "circle"){
        let c:Circle = s;
    }
}

function fail(s: Shapes){
    if(isShape(s)){
        succeed(s);
    }
}

```

**Comment by ahejlsberg** [comment]
Fix now available in #11198.
